### PR TITLE
CLDC-2316 Bulk upload sales 2023

### PR DESCRIPTION
**Pull request by asmega**
# Context

- https://digital.dclg.gov.uk/jira/browse/CLDC-2316
- Implement bulk upload sales for new collection year 2023
- This is a first pass implementation and will probably have some bugs in it and we can address over time

# Changes

- Add CSV parser for sales 2023 to handle CSV structure
- Tweak collection window validation so error now contextual to year selected for upload
- Handle arbitrary ordering of CSV columns
- Fix ordering of errors in report and now ordered by cell
- Added `Upload your file again` link styled as button on error report to match lettings experience
- Update tooling to convert logs to 2023 csv rows with support for random column ordering

# Known issues

- There seem to be some issues with how UPRN is handled if the UPRN cannot be validated.
- For the above I think there is dependency on https://github.com/communitiesuk/submit-social-housing-lettings-and-sales-data/pull/1570 which should clear any errored fields so users can continue to create logs and fix within the service

### CODE REVIEW

File: app/services/bulk_upload/sales/year2023/csv_parser.rb [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+  def column_for_field(field)
+    cols[field_numbers.find_index(field) + col_offset]
+  end
+
+private
+
+  def default_field_numbers
+    [6, 3, 4, 5, nil, 28, 30, 38, 47, 51, 55, 59, 31, 39, 48, 52, 56, 60, 37, 46, 50, 54, 58, 35, 43, 49, 53, 57, 61, 32, 33, 78, 80, 79, 81, 83, 84, nil, 62, 66, 64, 65, 63, 67, 69, 70, 68, 76, 77, 16, 17, 18, 26, 24, 25, 27, 8, 91, 95, 96, 97, 92, 93, 94, 98, 100, 101, 103, 104, 106, 110, 111, 112, 113, 114, 9, 116, 117, 118, 120, 124, 125, 126, 10, 11, nil, 127, 129, 133, 134, 135, 1, 2, nil, 73, nil, 75, 107, 108, 121, 122, 130, 131, 82, 109, 123, 132, 115, 15, 86, 87, 29, 7, 12, 13, 14, 36, 44, 45, 88, 89, 102, 105, 119, 128, 19, 20, 21, 22, 23, 34, 40, 41, 42, 71, 72, 74, 85, 90, 99].map do |h|
+      if h.present? && h.to_s.match?(/^[0-9]+$/)
```
Comment (natdeanlewissoftwire) [inline app/services/bulk_upload/sales/year2023/csv_parser.rb]: why is this called h?

File: app/services/bulk_upload/sales/year2023/csv_parser.rb [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+  def row_sep
+    "\n"
+  end
+
+  def normalised_string
+    return @normalised_string if @normalised_string
+
+    @normalised_string = File.read(path, encoding: "bom|utf-8")
+    @normalised_string.gsub!("\r\n", "\n")
```
Comment (natdeanlewissoftwire) [inline app/services/bulk_upload/sales/year2023/csv_parser.rb:86]: could use your row_sep method here

File: spec/support/bulk_upload/sales_log_to_csv.rb [hunk 3]
```diff
@@ -28,6 +37,34 @@ def default_2022_field_numbers_row(seed: nil)
...
+  end
+
+  def to_2023_row
+    to_2022_row + [
+      log.uprn,
+      log.address_line1,
+      log.address_line2,
+      log.town_or_city,
+      log.county,
+      nil, # mistake and now field is left blank
```
Comment (natdeanlewissoftwire) [inline spec/support/bulk_upload/sales_log_to_csv.rb]: i'm not sure what this comment means - if it's needed could it be a bit clearer?

**Review by natdeanlewissoftwire — APPROVED** [review]
Looks good to me technically, couple of questions but happy to approve at this point :) Seems like one worth extra thoroughly PO reviewing to try to find some of the possible bugs ASAP

File: app/services/bulk_upload/sales/year2023/csv_parser.rb (hunk 1 above)
Comment (jamesrose) [reply app/services/bulk_upload/sales/year2023/csv_parser.rb]: nit: nil check not required here

Comment (asmega) [reply app/services/bulk_upload/sales/year2023/csv_parser.rb]: > why is this called h?

it was a probably a hash at some point? so just abbreviated to `h`. i will change to something else, maybe `number`?

File: spec/support/bulk_upload/sales_log_to_csv.rb (hunk 3 above)
Comment (asmega) [reply spec/support/bulk_upload/sales_log_to_csv.rb]: theres a mistake in the template therefore ignored. i've updated the comment to reflect this

File: app/services/bulk_upload/sales/year2023/csv_parser.rb (hunk 1 above)
Comment (asmega) [reply app/services/bulk_upload/sales/year2023/csv_parser.rb]: i've now updated this

File: app/services/bulk_upload/sales/year2023/csv_parser.rb (hunk 2 above)
Comment (asmega) [reply app/services/bulk_upload/sales/year2023/csv_parser.rb:86]: i don't feel there's much benefit to re-using that method here, so i'm gonna leave this as is

File: app/services/bulk_upload/sales/year2023/csv_parser.rb (hunk 1 above)
Comment (natdeanlewissoftwire) [reply app/services/bulk_upload/sales/year2023/csv_parser.rb]: ah i see, thanks

File: spec/support/bulk_upload/sales_log_to_csv.rb (hunk 3 above)
Comment (natdeanlewissoftwire) [reply spec/support/bulk_upload/sales_log_to_csv.rb]: Gotcha, that makes sense

**Pull request merged**